src/compression/b44/mod.rs

```diff
@@ -0,0 +1,924 @@
+
+// Unpack a 14-byte block into 4 by 4 16-bit pixels.
+fn unpack14(b: &[u8], s: &mut [u16; 16]) {
+    // TODO: b slice should be &mut [u8; 14], but rust doesn't support sized slice.
+    assert_eq!(b.len(), 14);
+    assert_ne!(b[2], 0xfc);
+
+    s[0] = ((b_u32(&b, 0) << 8) | b_u32(&b, 1)) as u16;
+
+    let shift = (b_u32(&b, 2) >> 2);
+    let bias = 0x20 << shift;
+
+    s[4] = (s[0] as u32 + ((((b_u32(&b, 2) << 4) | (b_u32(&b, 3) >> 4)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[8] = (s[4] as u32 + ((((b_u32(&b, 3) << 2) | (b_u32(&b, 4) >> 6)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[12] = (s[8] as u32 + ((b_u32(&b, 4) & SIX_BITS) << shift) - bias) as u16;
+
+    s[1] = (s[0] as u32 + ((b_u32(&b, 5) >> 2) << shift) - bias) as u16;
+    s[5] = (s[4] as u32 + ((((b_u32(&b, 5) << 4) | (b_u32(&b, 6) >> 4)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[9] = (s[8] as u32 + ((((b_u32(&b, 6) << 2) | (b_u32(&b, 7) >> 6)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[13] = (s[12] as u32 + ((b_u32(&b, 7) & SIX_BITS) << shift) - bias) as u16;
+
+    s[2] = (s[1] as u32 + ((b_u32(&b, 8) >> 2) << shift) - bias) as u16;
+    s[6] = (s[5] as u32 + ((((b_u32(&b, 8) << 4) | (b_u32(&b, 9) >> 4)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[10] = (s[9] as u32 + ((((b_u32(&b, 9) << 2) | (b_u32(&b, 10) >> 6)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[14] = (s[13] as u32 + ((b_u32(&b, 10) & SIX_BITS) << shift) - bias) as u16;
+
+    s[3] = (s[2] as u32 + ((b_u32(&b, 11) >> 2) << shift) - bias) as u16;
+    s[7] = (s[6] as u32 + ((((b_u32(&b, 11) << 4) | (b_u32(&b, 12) >> 4)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[11] = (s[10] as u32 + ((((b_u32(&b, 12) << 2) | (b_u32(&b, 13) >> 6)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[15] = (s[14] as u32 + ((b_u32(&b, 13) & SIX_BITS) << shift) - bias) as u16;
+
+    for i in 0..16 {
+        if (s[i] & 0x8000) != 0 {
+            s[i] &= 0x7fff;
+        } else {
+            s[i] = !s[i];
+        }
+    }
+}
+
+// Unpack a 3-byte block into 4 by 4 identical 16-bit pixels.
+fn unpack3(b: &[u8], s: &mut [u16; 16]) {
+    // TODO: b slice should be &mut [u8; 3], but rust doesn't support sized slice.
+    assert_eq!(b[2], 0xfc);
+
+    s[0] = (((b[0] as u32) << 8) | (b[1] as u32)) as u16;
+
+    if (s[0] & 0x8000) != 0 {
+        s[0] &= 0x7fff;
+    } else {
+        s[0] = !s[0];
+    }
+
+    for i in 0..16 {
+        s[i] = s[0];
+    }
+}
+
+#[derive(Debug)]
+struct ChannelData {
+    tmp_start_index: usize,
+    tmp_end_index: usize,
+
+    resolution: Vec2<usize>,
+    y_sampling: usize,
+    type_: SampleType,
+    quantize_linearly: bool,
+    samples_per_pixel: usize,
+}
+
+fn cpy_u16(src: &[u16], src_i: usize, dst: &mut [u16], dst_i: usize, n: usize) {
+    // assert_eq!(src.len(), dst.len());
+
+    // for i in 0..src.len() {
+    for i in 0..n {
+        dst[dst_i + i] = src[src_i + i];
+    }
+}
```

Comment:
This code and the place using it are kinda ugly. It's supposed to be a `memcpy` and I'm not sure this code is the clearer to write.

Reply:
Yes, seems like it is just a copy, but not from the beginning, but from some start index.

```rust
fn copy_u16(src: &[u16], src_i: usize, dst: &mut [u16], dst_i: usize, n: usize) {
    let source      = &    src[src_i .. src_i + n]; // make a slice from start index to end
    let destination = &mut dst[dst_i .. dst_i + n]; // make a mutable slice from start index to end

    destination.copy_from_slice(source); // copy all slice values

   // for educational purposes, here's how I would do that by hand, if the method above did not exist:
    for (src_elem, dst_elem) in source.iter().zip(destination.iter()) {  // iterate through both slices simultaneously
      *dst_elem = *src_elem; // copy each element
    }
}
```

Considering that the actual copying is a single function call, I would probably see if this function can be removed completely, and work with the slices directly. Maybe the whole scr_i and n variables are not necessary at all, if you work with slices all the time, instead of using indices. At least that's what I usually try to do.

Reply:
Thanks, the `copy_to_slice()` is the important part.

I think a function make the code simpler to read (I will rename the function, of course). Changing this:

```rust
cpy_u16(&s, 0, &mut tmp_u16_buffer, row0, sample_count);
cpy_u16(&s, 4, &mut tmp_u16_buffer, row1, sample_count);
cpy_u16(&s, 8, &mut tmp_u16_buffer, row2, sample_count);
cpy_u16(&s, 12, &mut tmp_u16_buffer, row3, sample_count);
```
To this:

```rust
tmp_u16_buffer[row0..(row0+sample_count)].copy_from_slice(&s[0..sample_count]);
tmp_u16_buffer[row1..(row1+sample_count)].copy_from_slice(&s[4..(4+sample_count)]);
tmp_u16_buffer[row2..(row2+sample_count)].copy_from_slice(&s[8..(8+sample_count)]);
tmp_u16_buffer[row3..(row3+sample_count)].copy_from_slice(&s[12..(12+sample_count)]);
```

Seems more confusing to me.

`cpy_u16()` is used at some other places.

Reply:
As mentioned in my other suggestions, I would do it very simple:

```rust
row0.copy_from_slice(&s[0 .. 4]);
row1.copy_from_slice(&s[4 .. 8]); // or [ 4 .. 4 + sample_count ] ?
row2.copy_from_slice(&s[8 .. 12]);
row3.copy_from_slice(&s[12 .. 16]);
```

or with a loop? might be more complicated, but maybe you get the idea:
```rust
let sources = s.chunks_exact(4).map(|s| &s[..sample_count]);
let rows = tmp_buffer.chunks_exact(row_length);
for (row, source) in rows.zip(sources) { 
    row.copy_from_slice(source); 
}
```

By trying to work with slices only, never really using indices in the first place. Might be worth a try.

Reply:
> And try to have the slices, no indices, throughout the whole algorithm.

Ah, OK, make sense. I will try to see how I can do that.